dao/rhc_connection_dao.go

```diff
@@ -0,0 +1,192 @@
+package dao
+
+import (
+	"fmt"
+
+	m "github.com/RedHatInsights/sources-api-go/model"
+	"github.com/RedHatInsights/sources-api-go/util"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+type RhcConnectionDaoImpl struct{}
+
+func (s *RhcConnectionDaoImpl) List(limit, offset int, filters []util.Filter) ([]m.RhcConnection, int64, error) {
```

Comment:
Actually one more thing - I see you used the many2many thing - we may be able to preload the associations!

`DB.Model(&m.Source{}).Preload("RhcConnections").Where("tenant_id = ?", s.TenantID).Scan(...)` would work maybe.

Reply:
Also another thought: if we added tenancy to the join table we wouldn't have to do the double join on the sources table. What do you think of that?

Reply:
> Actually one more thing - I see you used the many2many thing - we may be able to preload the associations!

I was avoiding to use `.Preload` because GORM just throws as many `SELECT * FROM`s as `.Preload` calls you may have, to then match everything in memory[citation needed].

> Also another thought: if we added tenancy to the join table we wouldn't have to do the double join on the sources table. 

I foresee complications when syncing data among the tables. First, we would need to create a new row in the join table every time a new source is created. That would imply leaving the `rhc_connection_id` column as `NULL`. Then, every time we create a new connection, we would have to update that row to include the `rhc_connection_id`.

I guess that could work, but I'm debating on whether to use the following and call it a day:

```sql
SELECT "rhc_connections".*, STRING_AGG(CAST ("src"."id" AS text), ',')
FROM "rhc_connections"
INNER JOIN "source_rhc_connections" "jt" ON "rhc_connections"."id" = "jt"."rhc_connection_id"
INNER JOIN "sources" "src" ON "jt"."source_id" = "src.id" WHERE "src"."tenant_id" = ?
GROUP BY "rhc_connections"."id" LIMIT 100
```

Essentially this query brings all the connections and the aggregated source ids for a given tenant. This will make me have to pseudo manually scan the rows:

```go
rows, err := /* blablabla query */.Rows()
if err != nil { return err }

for rows.Next() {
    var rhcConnection m.RhcConnection
    var sourceIds map[string]interface{}

    err = DB.Scan(rows, &rhcConnection)
    if err != nil { return err }

    err = DB.Scan(rows, &sourceIds)
    if err != nil { return err }
    
    // for each source ID create a new source object and append it to the rhcConnection.Sources slice
```

It is somewhat manual but I kind of like it. In order to produce the above query the following Go code is needed:

```go
	query := DB.
		Debug().
		Model(&m.RhcConnection{}).
		Select(`"rhc_connections".*, STRING_AGG(CAST ("src"."id" AS TEXT), ',') AS "source_ids"`).
		Joins(`INNER JOIN "source_rhc_connections" "jt" ON "rhc_connections"."id" = "jt"."rhc_connection_id"`).
		Joins(`INNER JOIN "sources" "src" ON "jt"."source_id" = "src"."id"`).
		Where(`"src"."tenant_id" = ?`, s.TenantID).
		Group(`"rhc_connections"."id"`).
		Limit(limit).
		Offset(offset)
```

What do you think?

Reply:
>First, we would need to create a new row in the join table every time a new source is created. 

Yeah that is pretty standard, and once we add the foreign key it will be fine. We don't need to leave any records NULL, at least I don't see where/why. For resource creation it's always going to be:
1. `DB.FirstOrCreate(&m.RhcConnection{RhcID: "asdfasdfaf")`
2. `DB.Create(&m.SourceRhcConnection{SourceID: <srcid from payload>, RhcConnectionID: <rhcid>})`

Then for listing we'll have to get fancy with the SQL - but if we add tenancy to the join talbe we won't have to double-join things in anymore at least. It'll be something like:
```sql
SELECT DISTINCT rhc_connections.*
FROM source_rhc_connections 
INNER JOIN rhc_connections ON rhc_connections.id = source_rhc_connections.rhc_connection_id
WHERE tenant_id = 1
```
This should be pretty easy to emulate with GORM at least.

Reply:
> Yeah that is pretty standard, and once we add the foreign key it will be fine. We don't need to leave any records NULL, at least I don't see where/why. For resource creation it's always going to be:

You are right: when creating the connection we already will be able to create the row in the join table, since all the required ids will come in the request —source id, tenant id and the connection id that we will grab from the freshly created one—.

All right, suggestion implemented.